### PR TITLE
feat(hyd): fluid dynamics in reservoirs

### PR DESCRIPTION
**Pull request by crocket63**
Fixes #[issue_no]

## Summary of Changes

An simplified oscillation/viscous model will simulate fluid movement in hydraulic reservoirs.

Lateral/longitudinal accelerations might impact fluid level seen by the gauge.
Flying close to 0Gs might have strange impacts on gauge reading

Reservoirs have a small negative G cavity that can hold fluid for the pumps for around 20s. After that time pumps will cavitate and hydraulics are lost.

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Note we don't have real references for those extreme behaviors, so everything can be tuned later. We only need to check that it behaves "incorrectly" as we want it to be.

-Fluid level can be seen moving a bit when plane accelerations occur (left/right, accelerate/brake on ground).
-Negative Gs cause erratic hyd level readings, but hydraulics are still working
-Long period in negative G (inverted flight) should cause hydraulic failure (in around 20s)

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by crocket63** [comment]
Let debug print in here for other tweaks after first feedbacks

File: src/systems/systems/src/hydraulic/mod.rs [hunk 1]
```diff
@@ -1622,6 +1625,185 @@ impl SimulationElement for LeakMeasurementValve {
...
+
+    fn new() -> Self {
+        Self {
+            reference_point_cg: Vector3::default(),
+            fluid_cg_position: Vector3::new(0., -0.2, 0.),
+            fluid_cg_speed: Vector3::default(),
+
+            virtual_mass: Mass::new::<kilogram>(100.),
+            spring: SpringPhysics::new(),
+            anisotropic_damping_constant: Vector3::new(25., 20., 25.),
```
Comment (BlueberryKing) [inline src/systems/systems/src/hydraulic/mod.rs:1694]: Do most of these constants not depend on the geometry of the container (e.g the reservoir)? If so, they should probably be passed in or maybe we could explicitly designate this struct to be `ReservoirFluidPhysics`

File: src/systems/systems/src/hydraulic/mod.rs [hunk 2]
```diff
@@ -1622,6 +1625,185 @@ impl SimulationElement for LeakMeasurementValve {
...
+
+        // println!(
+        //     "current_fluid_cg {:.2} {:.2} {:.2} GTRAP EMPTY {:?} GAUGE {:.2}  USABLE {:.2}",
+        //     self.fluid_cg_position[0],
+        //     self.fluid_cg_position[1],
+        //     self.fluid_cg_position[2],
+        //     self.g_trap_is_empty.output(),
+        //     self.gauge_modifier().get::<ratio>(),
+        //     self.usable_level_modifier().get::<ratio>()
+        // );
```
Comment (BlueberryKing) [inline src/systems/systems/src/hydraulic/mod.rs:1719]:
```suggestion
```

File: src/systems/systems/src/hydraulic/mod.rs [hunk 3]
```diff
@@ -1622,6 +1625,185 @@ impl SimulationElement for LeakMeasurementValve {
...
+    fn new() -> Self {
+        Self {
+            k_constant: 5000.,
+            damping_constant: 500.,
+
+            last_length: 0.,
+        }
+    }
+
+    fn update_force_nm(
```
Comment (BlueberryKing) [inline src/systems/systems/src/hydraulic/mod.rs:1644]: What does the `nm` here stand for? Newton-meters? Newton and metres?

File: src/systems/systems/src/hydraulic/mod.rs [hunk 4]
```diff
@@ -1622,6 +1625,185 @@ impl SimulationElement for LeakMeasurementValve {
...
+
+        self.fluid_cg_speed += acceleration * context.delta_as_secs_f64();
+
+        self.fluid_cg_position += self.fluid_cg_speed * context.delta_as_secs_f64();
+    }
+
+    // Returns a coefficient of the actual level that will be seen by the gauge.
+    // Example: 0.5 means gauge reads half the actual level of the reservoir
+    fn gauge_modifier(&self) -> Ratio {
+        const LATERAL_BREAKPOINTS: [f64; 6] = [-1., -0.2, 0., 0.2, 0.4, 1.];
```
Comment (BlueberryKing) [inline src/systems/systems/src/hydraulic/mod.rs:1748]: Why are these breakpoints not symmetrical? Does this indicate that the gauge is not centered laterally? Also, why are the breakpoints between -1 and 1? The CG position is not normalized from what I can tell.

File: src/systems/systems/src/hydraulic/mod.rs (hunk 1 above)
Comment (crocket63) [reply src/systems/systems/src/hydraulic/mod.rs:1694]: This is really macro simulated and not using physical dimensions at all. So this is more like general constants that seem to work with fluid in reservoir in general, not dedicated to a A320 reservoir.

All the maps used however are what we'd change on a reservoir/plane basis I'd say.

File: src/systems/systems/src/hydraulic/mod.rs (hunk 3 above)
Comment (crocket63) [reply src/systems/systems/src/hydraulic/mod.rs:1644]: Newton meters here, but as nothing here has true physical sense....  Not sure I should even mention units

File: src/systems/systems/src/hydraulic/mod.rs (hunk 4 above)
Comment (crocket63) [reply src/systems/systems/src/hydraulic/mod.rs:1748]: Exactly: adjusting the maps you can account for gauge sensor position.  If it's on the right you'd get more level on right accelerations less on left.
CG position is not normalized indeed, 1 is just chosen as "if you get 1 you are already dead and reservoir is not a problem anymore"

File: src/systems/systems/src/hydraulic/mod.rs [hunk 5]
```diff
@@ -1622,6 +1625,185 @@ impl SimulationElement for LeakMeasurementValve {
     }
 }
 
+struct SpringPhysics {
+    k_constant: f64,
+    damping_constant: f64,
```
Comment (davidwalschots) [inline src/systems/systems/src/hydraulic/mod.rs:1630]: *suggestion**: Use the `const NAME: TYPE = VALUE;` syntax instead.

File: src/systems/systems/src/hydraulic/mod.rs [hunk 6]
```diff
@@ -1622,6 +1625,185 @@ impl SimulationElement for LeakMeasurementValve {
...
+            &LONGITUDINAL_BREAKPOINTS,
+            &LONGITUDINAL_MAP,
+            self.fluid_cg_position[2],
+        );
+
+        let vertical_ratio = interpolation(
+            &VERTICAL_BREAKPOINTS,
+            &VERTICAL_MAP,
+            self.fluid_cg_position[1],
+        );
```
Comment (davidwalschots) [inline src/systems/systems/src/hydraulic/mod.rs:1773]: Out of curiousity. Is the order lateral, vertical, longitudinal in the `Vector` common practice? For me (as a total n00b on the subject) it feels more natural to think of them in the order lateral, longitudinal and vertical.

**side note**: `logitudinal_ratio` has a typo lo*n*gitudinal_ratio.

File: src/systems/systems/src/hydraulic/mod.rs [hunk 7]
```diff
@@ -1622,6 +1625,185 @@ impl SimulationElement for LeakMeasurementValve {
...
+    // Returns a coefficient of the actual level that will be seen by the gauge.
+    // Example: 0.5 means gauge reads half the actual level of the reservoir
+    fn gauge_modifier(&self) -> Ratio {
+        const LATERAL_BREAKPOINTS: [f64; 6] = [-1., -0.2, 0., 0.2, 0.4, 1.];
+        const LATERAL_MAP: [f64; 6] = [0.2, 0.95, 1., 1.05, 1.2, 1.3];
+
+        const LONGITUDINAL_BREAKPOINTS: [f64; 6] = [-1., -0.1, 0., 0.1, 0.4, 1.];
+        const LONGITUDINAL_MAP: [f64; 6] = [0.2, 0.98, 1., 0.98, 0.2, 0.2];
+
+        const VERTICAL_BREAKPOINTS: [f64; 6] = [-1., -0.1, 0., 0.1, 0.2, 1.];
```
Comment (davidwalschots) [inline src/systems/systems/src/hydraulic/mod.rs:1754]: **suggestion**: Consider moving the consts closer to the calculation, e.g.:

```rust
const VERTICAL_BREAKPOINTS: [f64; 6] = [-1., -0.1, 0., 0.1, 0.2, 1.];
const VERTICAL_MAP: [f64; 6] = [1., 1., 0.7, 0.9, 1.2, 1.5];

let vertical_ratio = interpolation(
    &VERTICAL_BREAKPOINTS,
    &VERTICAL_MAP,
    self.fluid_cg_position[1],
);
```

It's a small thing, but it slightly reduces "context switching" as your eyes glide over the different axes. 😀

File: src/systems/systems/src/hydraulic/mod.rs [hunk 8]
```diff
@@ -1622,6 +1625,185 @@ impl SimulationElement for LeakMeasurementValve {
...
+        const LONGITUDINAL_BREAKPOINTS: [f64; 6] = [-1., -0.1, 0., 0.1, 0.4, 1.];
+        const LONGITUDINAL_MAP: [f64; 6] = [0.2, 1., 1., 1., 0.2, 0.2];
+
+        let lateral_ratio = interpolation(
+            &LATERAL_BREAKPOINTS,
+            &LATERAL_MAP,
+            self.fluid_cg_position[0],
+        );
+
+        let logitudinal_ratio = interpolation(
```
Comment (davidwalschots) [inline src/systems/systems/src/hydraulic/mod.rs:1793]:
```suggestion
        let longitudinal_ratio = interpolation(
```

It seems multiline selection in PR reviewing is broken. So I can't make a suggestion which you can immediately commit. Sorry about that 😀.

**Review by davidwalschots — COMMENTED** [review]
Some minor comments. Cool stuff!

File: src/systems/systems/src/hydraulic/mod.rs (hunk 6 above)
Comment (lukecologne) [reply src/systems/systems/src/hydraulic/mod.rs:1773]: The way I know it, a standard coordinate system for aircraft is x,y,z - longitudinal, lateral, vertical, with a right-handed coordinate system: 
grafik
(This is from the book Flugregelung)

MSFS is a bit odd with it's coordinate system, and I think it is also left-handed.

File: src/systems/systems/src/hydraulic/mod.rs (hunk 7 above)
Comment (crocket63) [reply src/systems/systems/src/hydraulic/mod.rs:1754]: Good point :)

File: src/systems/systems/src/hydraulic/mod.rs (hunk 6 above)
Comment (crocket63) [reply src/systems/systems/src/hydraulic/mod.rs:1898]: Yeah I'm getting used to the msfs coord indeed :p  Y being down/up

**Review by BlueberryKing — APPROVED** [review]

**Review by davidwalschots — APPROVED** [review]

File: .github/CHANGELOG.md [hunk 9]
```diff
@@ -61,6 +61,7 @@
 1. [HYD] Randomised per actuator flow restrictions at plane init - @Crocket63 (crocket)
 1. [MCDU] Hide stored elements on A/C Status when there are none - @tracernz (Mike)
 1. [FMGC] Fix ident for CD legs - @tracernz (Mike)
+1. [HYD] Accelerations impacts hydraulic reservoir fluid movement - @Crocket63 (crocket)
```
Comment (BlueberryKing) [inline .github/CHANGELOG.md:64]: Sorry for not noticing this earlier, but could you move this entry to the bottom?

**Comment by EuanM199** [comment]
Quality Assurance Trainee Report

Discord   :  Euan#7980
object of testing: #7393
Tier of Testing: 2
Date: 19/09/2022

Testing Process: 

- [x] Fluid level can be seen moving a bit when plane accelerations occur (left/right, accelerate/brake on ground).
- [x] -Negative Gs cause erratic hyd level readings, but hydraulics are still working
- [x] Long period in negative G (inverted flight) should cause hydraulic failure (in around 20s) 

Media:

On the Ground:
https://user-images.githubusercontent.com/67420304/191013051-83f94182-79bd-4422-a0a0-1ba7218684f3.mp4

During Takeoff:
https://user-images.githubusercontent.com/67420304/191024712-4bcf2c5d-a7df-4e75-bc91-8ba8174e05d8.mp4

During Inversion and Long Period of Negitive Gs

https://user-images.githubusercontent.com/67420304/191025042-9b926214-c3dc-4ad3-9ac0-ac85a94b147b.mp4


Negatives:       
N/A

Testing Results: 
Passed

File: .github/CHANGELOG.md (hunk 9 above)
Comment (crocket63) [reply .github/CHANGELOG.md]: Done

**Pull request merged**